File: core/java/com/android/internal/widget/remotecompose/core/operations/utilities/IntMap.java
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.internal.widget.remotecompose.core.operations.utilities;

import java.util.ArrayList;
import java.util.Arrays;

public class IntMap<T> {

    private static final int DEFAULT_CAPACITY = 16;
    private static final float LOAD_FACTOR = 0.75f;
    private static final int NOT_PRESENT = Integer.MIN_VALUE;
    private int[] mKeys;
    private ArrayList<T> mValues;
    int mSize;

    public IntMap() {
        mKeys = new int[DEFAULT_CAPACITY];
        Arrays.fill(mKeys, NOT_PRESENT);
        mValues = new ArrayList<T>(DEFAULT_CAPACITY);
        for (int i = 0; i < DEFAULT_CAPACITY; i++) {
            mValues.add(null);
        }
    }

    public void clear() {
        Arrays.fill(mKeys, NOT_PRESENT);
        mValues.clear();
        mSize = 0;
    }

    public T put(int key,  T value)  {
        if (key == NOT_PRESENT) throw new IllegalArgumentException("Key cannot be NOT_PRESENT");
        if (mSize > mKeys.length * LOAD_FACTOR) {
            resize();
        }
        return insert(key, value);
    }


    public  T get(int key) {
        int index = findKey(key);
        if (index == -1) {
            return  null;
        } else
            return mValues.get(index);
    }

    public int size() {
        return mSize;
    }

    private  T insert(int key, T value) {
        int index = hash(key) % mKeys.length;
        while (mKeys[index] != NOT_PRESENT && mKeys[index] != key) {
            index = (index + 1) % mKeys.length;
        }
        T oldValue =  null;
        if (mKeys[index] == NOT_PRESENT) {
            mSize++;
        } else {
            oldValue = mValues.get(index);
        }
        mKeys[index] = key;
        mValues.set(index, value);
        return oldValue;
    }

    private  int findKey(int key) {
        int index = hash(key) % mKeys.length;
        while (mKeys[index] != NOT_PRESENT) {
            if (mKeys[index] == key) {
                return index;
            }
            index = (index + 1) % mKeys.length;
        }
        return -1;
    }

    private  int hash(int key) {
        return key;
    }

    private   void resize() {
        int[] oldKeys = mKeys;
        ArrayList<T> oldValues = mValues;
        mKeys = new int[(oldKeys.length * 2)];
        for (int i = 0; i < mKeys.length; i++) {
            mKeys[i] = NOT_PRESENT;
        }
        mValues = new ArrayList<T>(oldKeys.length * 2);
        for (int i = 0; i < oldKeys.length * 2; i++) {
            mValues.add(null);
        }
        mSize = 0;
        for (int i = 0; i < oldKeys.length; i++) {
            if (oldKeys[i] != NOT_PRESENT) {
                put(oldKeys[i], oldValues.get(i));
            }
        }
    }
}

File: services/permission/java/com/android/server/permission/access/immutable/IntMap.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.server.permission.access.immutable

import android.util.SparseArray

/** Immutable map with index-based access and [Int] keys. */
sealed class IntMap<T>(internal val array: SparseArray<T>) : Immutable<MutableIntMap<T>> {
    val size: Int
        get() = array.size()

    fun isEmpty(): Boolean = array.size() == 0

    operator fun contains(key: Int): Boolean = array.contains(key)

    operator fun get(key: Int): T? = array.get(key)

    fun indexOfKey(key: Int): Int = array.indexOfKey(key)

    fun keyAt(index: Int): Int = array.keyAt(index)

    fun valueAt(index: Int): T = array.valueAt(index)

    override fun toMutable(): MutableIntMap<T> = MutableIntMap(this)

    override fun toString(): String = array.toString()
}

/** Mutable map with index-based access and [Int] keys. */
class MutableIntMap<T>(array: SparseArray<T> = SparseArray()) : IntMap<T>(array) {
    constructor(intMap: IntMap<T>) : this(intMap.array.clone())

    fun put(key: Int, value: T): T? = array.putReturnOld(key, value)

    fun remove(key: Int): T? = array.removeReturnOld(key).also { array.gc() }

    fun clear() {
        array.clear()
    }

    fun putAt(index: Int, value: T): T = array.setValueAtReturnOld(index, value)

    fun removeAt(index: Int): T = array.removeAtReturnOld(index).also { array.gc() }
}

internal fun <T> SparseArray<T>.putReturnOld(key: Int, value: T): T? {
    val index = indexOfKey(key)
    return if (index >= 0) {
        val oldValue = valueAt(index)
        setValueAt(index, value)
        oldValue
    } else {
        put(key, value)
        null
    }
}

// SparseArray.removeReturnOld() is @hide, so a backup once we move to APIs.
@Suppress("EXTENSION_SHADOWED_BY_MEMBER")
internal fun <T> SparseArray<T>.removeReturnOld(key: Int): T? {
    val index = indexOfKey(key)
    return if (index >= 0) {
        val oldValue = valueAt(index)
        removeAt(index)
        oldValue
    } else {
        null
    }
}

internal fun <T> SparseArray<T>.setValueAtReturnOld(index: Int, value: T): T {
    val oldValue = valueAt(index)
    setValueAt(index, value)
    return oldValue
}

internal fun <T> SparseArray<T>.removeAtReturnOld(index: Int): T {
    val oldValue = valueAt(index)
    removeAt(index)
    return oldValue
}

internal fun <T> SparseArray<T>.gc() {
    size()
}
